Engine/Component/ComponentBillboard.cpp

```diff
@@ -0,0 +1,235 @@
+#include "ResourceManagement/ResourcesDB/CoreResources.h"
+
+ComponentBillboard::ComponentBillboard() : Component(nullptr, ComponentType::BILLBOARD)
+{
+	InitData();
+}
+
+ComponentBillboard::ComponentBillboard(GameObject * _owner) : Component(owner, ComponentType::BILLBOARD)
+{
+	InitData();
+}
+
+ComponentBillboard::~ComponentBillboard()
+{
+	if (vbo != 0)
+	{
+		glDeleteBuffers(1, &vbo);
+		glDeleteBuffers(1, &ebo);
+		glDeleteVertexArrays(1, &vao);
+	}
+}
+
+void ComponentBillboard::InitData()
+{
+	ChangeTexture(static_cast<uint32_t>(CoreResource::BILLBOARD_DEFAULT_TEXTURE));
+
+	float vertices[20] =
+	{
+			0.5f,  0.5f, 0.0f,		1.0f, 1.0f,
+			0.5f, -0.5f, 0.0f,		1.0f, 0.0f,
+		   -0.5f, -0.5f, 0.0f,		0.0f, 0.0f,
+		   -0.5f,  0.5f, 0.0f,		0.0f, 1.0f
+	};
+	unsigned int indices[6] =
+	{
+		0, 1, 3,
+		1, 2, 3
+	};
+
+	glGenVertexArrays(1, &vao);
+	glGenBuffers(1, &vbo);
+	glGenBuffers(1, &ebo);
+
+	glBindVertexArray(vao);
+
+	glBindBuffer(GL_ARRAY_BUFFER, vbo);
+	glBufferData(GL_ARRAY_BUFFER, sizeof(vertices), vertices, GL_STATIC_DRAW);
+
+	glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, ebo);
+	glBufferData(GL_ELEMENT_ARRAY_BUFFER, sizeof(indices), indices, GL_STATIC_DRAW);
+
+	glVertexAttribPointer(0, 3, GL_FLOAT, GL_FALSE, 5 * sizeof(float), (void*)0);
+	glEnableVertexAttribArray(0);
+
+	glVertexAttribPointer(1, 2, GL_FLOAT, GL_FALSE, 5 * sizeof(float), (void*)(3 * sizeof(float)));
+	glEnableVertexAttribArray(1);
+
+	glBindBuffer(GL_ARRAY_BUFFER, 0);
+	glBindVertexArray(0);
+}
+
+void ComponentBillboard::SwitchFrame()
+{
+	time_since_start += App->time->delta_time;
+	APP_LOG_INFO("%.1f", time_since_start);
+
+	if (time_since_start * sheet_speed >= 1)
+	{
+		current_sprite_x += 1;
+
+		if ((int)current_sprite_x >= x_tiles) {
+			current_sprite_y--;
+			current_sprite_x = 0;
+		}
+
+		if ((int)current_sprite_y <= 0) {
+			current_sprite_y = y_tiles;
+		}
+		time_since_start = 0.f;
+	}
+
+
+}
+
+void ComponentBillboard::ChangeBillboardType(ComponentBillboard::AlignmentType _alignment_type)
+{
+	alignment_type = _alignment_type;
+
+	if (alignment_type == SPRITESHEET)
+		is_spritesheet = true;
+	else
+		is_spritesheet = false;
+}
+
+void ComponentBillboard::Render(const float3& position)
+{
+	GLuint shader_program = App->program->GetShaderProgramId("Billboard");
+	glUseProgram(shader_program);
+
+	int n;
+	glGetProgramStageiv(shader_program, GL_VERTEX_SHADER, GL_ACTIVE_SUBROUTINE_UNIFORM_LOCATIONS, &n);
+	unsigned* subroutines_indices = new unsigned[n];
+
+	//Subroutine functions
+	GLuint viewpoint_subroutine = glGetSubroutineIndex(shader_program, GL_VERTEX_SHADER, "view_point_alignment");
+	GLuint crossed_subroutine = glGetSubroutineIndex(shader_program, GL_VERTEX_SHADER, "crossed_alignment");
+	GLuint axial_subroutine = glGetSubroutineIndex(shader_program, GL_VERTEX_SHADER, "axial_alignment");
+
+	//Subroutine uniform
+	int selector = glGetSubroutineUniformLocation(shader_program, GL_VERTEX_SHADER, "alignment_selector");
+
+	switch (alignment_type) {
```

Comment:
code style

Reply:
I don't find the problem in code style here, can you point it out?

Reply:
same as before